feat: support engine `sail`

keen85 opened the pull request:
fixes #35
This PR adds support for [sail](https://github.com/lakehq/sail) (that uses [apache datafusion](https://github.com/apache/datafusion)) as another processing engine.

mwc360 commented:
> @mwc360 would you be up for a review?
> 
> I mostly developed locally and only did some basic testing in Fabric. What is your approach to develop and test _in_ Fabric?

For quick testing, I'll build the WHL locally, drop in the Resources of a Notebook, then pip install from that location: `%pip install "builtin/lakebench-0.8.1-py3-none-any.whl[sail]"`

mwc360 commented:
Ok so I got it working in a Fabric Notebook. I'll request two changes before merging in:
1. I get the below error when initializing the engine after pip install. Restarting the kernel via (`import sys
sys.exit(0)`) fixes it. Since this doesn't have to be done for other engines, can you catch the error and return a message noting that session restart is required. If `self.is_fabric` is True, you can note that `notebookutils.session.restartPython()` or `import sys
sys.exit(0)` can be used.

```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[5], line 4
      1 from lakebench.engines import Sail
      2 from lakebench.benchmarks import TPCH
----> 4 engine = Sail(
      5     delta_abfss_schema_path = f"abfss://lakebench.........tpch_test"
      6 )
      8 benchmark = TPCH(
      9     engine=engine,
     10     scenario_name="SF1 - All Queries",
   (...)
     14     query_list=['q1']
     15     )
     16 benchmark.run(mode="query")

File ~/jupyter-env/python3.11/lib/python3.11/site-packages/lakebench/engines/sail.py:46, in Sail.__init__(self, delta_abfss_schema_path, cost_per_vcore_hour)
     37 sail_server_hostname, sail_server_port = (
     38     self.sail_server.listening_address[0],
     39     self.sail_server.listening_address[1],
     40 )
     42 from pyspark.sql import SparkSession
     44 self.spark = SparkSession.builder.remote(
     45     f"sc://{sail_server_hostname}:{sail_server_port}"
---> 46 ).getOrCreate()
     47 self.spark.conf.set("spark.sql.warehouse.dir", delta_abfss_schema_path)
     49 self.delta_abfss_schema_path = delta_abfss_schema_path

File ~/jupyter-env/python3.11/lib/python3.11/site-packages/pyspark/sql/session.py:444, in getOrCreate(self)
    430 def getOrCreate(self) -> "SparkSession":
    431     """Gets an existing :class:`SparkSession` or, if there is no existing one, creates a
    432     new one based on the options set in this builder.
    433 
    434     .. versionadded:: 2.0.0
    435 
    436     .. versionchanged:: 3.4.0
    437         Supports Spark Connect.
    438 
    439     Returns
    440     -------
    441     :class:`SparkSession`
    442 
    443     Examples
--> 444     --------
    445     This method first checks whether there is a valid global default SparkSession, and if
    446     yes, return that one. If no valid global default SparkSession exists, the method
    447     creates a new SparkSession and assigns the newly created SparkSession as the global
    448     default.
    449 
    450     >>> s1 = SparkSession.builder.config("k1", "v1").getOrCreate()
    451     >>> s1.conf.get("k1") == "v1"
    452     True
    453 
    454     The configuration of the SparkSession can be changed afterwards
    455 
    456     >>> s1.conf.set("k1", "v1_new")
    457     >>> s1.conf.get("k1") == "v1_new"
    458     True
    459 
    460     In case an existing SparkSession is returned, the config options specified
    461     in this builder will be applied to the existing SparkSession.
    462 
    463     >>> s2 = SparkSession.builder.config("k2", "v2").getOrCreate()
    464     >>> s1.conf.get("k1") == s2.conf.get("k1") == "v1_new"
    465     True
    466     >>> s1.conf.get("k2") == s2.conf.get("k2") == "v2"
    467     True
    468     """
    469     opts = dict(self._options)
    471     if is_remote_only():

File ~/jupyter-env/python3.11/lib/python3.11/site-packages/pyspark/sql/connect/session.py:18
      1 #
      2 # Licensed to the Apache Software Foundation (ASF) under one or more
      3 # contributor license agreements.  See the NOTICE file distributed with
   (...)
     15 # limitations under the License.
     16 #
     17 import uuid
---> 18 from pyspark.sql.connect.utils import check_dependencies
     20 check_dependencies(__name__)
     22 import json

File ~/jupyter-env/python3.11/lib/python3.11/site-packages/pyspark/sql/connect/utils.py:21
     19 from pyspark.loose_version import LooseVersion
     20 from pyspark.sql.pandas.utils import require_minimum_pandas_version, require_minimum_pyarrow_version
---> 21 from pyspark.errors import PySparkImportError
     24 def check_dependencies(mod_name: str) -> None:
     25     if mod_name == "__main__" or mod_name == "pyspark.sql.connect.utils":

ImportError: cannot import name 'PySparkImportError' from 'pyspark.errors' (/home/trusted-service-user/jupyter-env/python3.11/lib/python3.11/site-packages/pyspark/errors/__init__.py)
```

2. Initializing the engine isn't idempotent. When initializing the engine a second time, it returns the following error:
```
OSError                                   Traceback (most recent call last)
Cell In[7], line 4
      1 from lakebench.engines import Sail
      2 from lakebench.benchmarks import TPCH
----> 4 engine = Sail(
      5     delta_abfss_schema_path = f"abfss://lakebench..........tpch_test"
      6 )
      8 benchmark = TPCH(
      9     engine=engine,
     10     scenario_name="SF1 - All Queries",
   (...)
     14     query_list=['q1']
     15     )
     16 benchmark.run(mode="query")

File ~/jupyter-env/python3.11/lib/python3.11/site-packages/lakebench/engines/sail.py:36, in Sail.__init__(self, delta_abfss_schema_path, cost_per_vcore_hour)
     33 from pysail.spark import SparkConnectServer
     35 self.sail_server = SparkConnectServer(port=50051)
---> 36 self.sail_server.start(background=True)
     37 sail_server_hostname, sail_server_port = (
     38     self.sail_server.listening_address[0],
     39     self.sail_server.listening_address[1],
     40 )
     42 from pyspark.sql import SparkSession

File ~/jupyter-env/python3.11/lib/python3.11/site-packages/pysail/spark/__init__.py:31, in SparkConnectServer.start(self, background)
     26 def start(self, *, background=True) -> None:
     27     """Start the server.
     28 
     29     :param background: Whether to start the server in a background thread.
     30     """
---> 31     self._inner.start(background=background)

OSError: Address already in use (os error 98)
```

Also - FYI, Sail on a 2 vcore machine seems to have OOM at SF1 on Q72, I'm guessing that you tested on a larger compute that didn't have the OOM.
```
Error during Query - q72... SparkConnectGrpcException: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "internal error: task 7316 panicked with message "byte array offset overflow""
	debug_error_string = "UNKNOWN:Error received from peer  {grpc_status:13, grpc_message:"internal error: task 7316 panicked with message \"byte array offset overflow\""}"
```

mwc360 commented:
Also - with the existing benchmarks, the `sail_lightweight` option is all that's needed. Does it make sense to remove the other sail option (with pyspark-connect)? I'd prefer to have one install option per engine.

keen85 commented:
Hi @mwc360 , thanks for the review.
That's some fair points.

1. concerning leightweight (`pyspark-client`) vs heavy (`pyspark[connect]`):
After reviewing [Sail's doc](https://docs.lakesail.com/sail/latest/introduction/getting-started/#package-installation) once more, I think that unfortunately we _need_ the heavy package, because it's a requirement for running the Sail SparkConnectServer.
2. concerning idempotence: if a Sail SparkConnectServer was started, and the refernce to it is lost (e.g by executing `engine = Sail()`) I did not find a way to reestablish a reference to the SparkConnectServer (in order to obtain it's status). Therefore, I was thinking about using a Singleton pattern making sure that several engine instances use the same SparkConnectServer and it is only started once.
3. concerning pointing out necessity of kernel restart: I realized, that the implementation of `BaseEngine.is_fabric` does not work in my workspace - `notebookutils.runtime.context['productType']` does not exist. I adjusted it and added throwing a conditional error message.

I haven't tested my changes yet properly; I will do so tomorrow.

mwc360 commented:
@keen85 - on the kernel restart, I'd take the Spark 4 reference out of there. It could be a bit misleading as the error occurs on a regular Python Notebook (no spark session). I'm guessing that error might occur because there could be some spark related binaries related to 3.5 in the runtime and connect/client target 4.0.

keen85 commented:
Thanks for chiming in here @linhr.
You are right, locally lightweight `pyspark-client` works fine for me, on my local machine.

Unfortunately it does not in Fabric

@mwc360 I did some more testing. It seems that after installing `pyspark-client==4.0.0` this breaks Microsofs pre-installed `sempy` in Fabric.

use this to reproduce:
```python
%pip install "pyspark-client==4.0.0"
import sys; sys.exit(0)
from sempy import fabric
```
```python
log_telemetry: sempy.relationships
log_telemetry: sempy.dependencies
log_telemetry: sempy.fabric
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[2], line 1
----> 1 from sempy import fabric

File ~/jupyter-env/python3.10/lib/python3.10/site-packages/sempy/__init__.py:18
     13 from ._utils._telemetry import log_telemetry
     16 __version__ = _version.get_versions()['version']
---> 18 _initialize_log(
     19     on_fabric=_on_fabric(),
     20     on_aiskill=_on_aiskill(),
     21     env=_get_environment(),
     22     notebook_workspace_id=get_notebook_workspace_id(),
     23     artifact_id=get_artifact_id(),
     24     artifact_type=_get_artifact_type(),
     25     fabric_run_id=_get_fabric_run_id(),
     26     root_activity_id=_get_root_activity_id()
     27 )
     29 # log telemetry
     30 log_telemetry(activity_name="sempy")

File ~/jupyter-env/python3.10/lib/python3.10/site-packages/sempy/_utils/_log.py:169, in _initialize_log(on_fabric, on_aiskill, env, notebook_workspace_id, artifact_id, artifact_type, fabric_run_id, root_activity_id)
    166         handler = logging.StreamHandler()
    167 else:
    168     # Running on Spark Cluster
--> 169     from synapse.ml.pymds.handler import SynapseHandler
    170     # Use the pymds decorator after it's aligned with the one we have here
    171     # from synapse.ml.pymds.synapse_logger import DecoratorJSONFormatter
    172     handler = SynapseHandler(MDS_LOG_TABLE, scrubbers=[])

File ~/jupyter-env/python3.10/lib/python3.10/site-packages/synapse/ml/pymds/__init__.py:1
----> 1 from .synapse_logger import get_mds_logger, get_mds_json_logger
      2 from .decorator import mds_log
      3 from ..internal_utils.version import VERSION   # type: ignore

File ~/jupyter-env/python3.10/lib/python3.10/site-packages/synapse/ml/pymds/synapse_logger.py:2
      1 import logging
----> 2 from .handler import SynapseHandler, LOGGER_ROOT_PREFIX
      3 from .scrubbers.scrubber import IScrub
      4 from synapse.ml.internal_utils.constants import ML_KUSTO_TABLE_NAME

File ~/jupyter-env/python3.10/lib/python3.10/site-packages/synapse/ml/pymds/handler.py:12
     10 from .scrubbers.scrubber import IScrub, JWTTokenScrubber, SmtpScrubber, InvalidLicenseScrubber, HiveRowExceptionScrubber, CreditCardScrubber, AppAttemptIdScrubber, SASSignatureScrubber
     11 from .internal_logger import log_debug
---> 12 from ..internal_utils.session_utils import get_fabric_context, is_pure_python_env, is_aikill_container
     13 from fluent import asyncsender as sender  # type: ignore
     14 from typing import Dict, Any, List, Optional

File ~/jupyter-env/python3.10/lib/python3.10/site-packages/synapse/ml/internal_utils/session_utils.py:35
     33 from urllib.parse import urlparse
     34 from .constants import SYNAPSE_CLUSTER_IDENTIFIER, SYNAPSE_CLUSTER_TYPE, SYNAPSE_TOKEN_SERVICE_ENDPOINT, TRIDENT_AISKILL_ENV, NOTEBOOKUTILS_RUNTIME_TYPE
---> 35 from pyspark import SparkContext
     36 from pyspark.sql import SparkSession
     38 logger = logging.getLogger(__name__)

ImportError: cannot import name 'SparkContext' from 'pyspark' (/home/trusted-service-user/jupyter-env/python3.10/lib/python3.10/site-packages/pyspark/__init__.py)
```

I think our options are:
- use `pyspark[connect]` (heavy)
- let users chose between `pyspark[connect]` (heavy) and `pyspark-client` (lightweight) and give a hint that in combination with Fabric only `pyspark[connect]` works
- use `pyspark-client` (lightweight); adjust `BaseEngine` so it does not use `sempy` any more (no more `FabricRestClient` 😣)
@mwc360 pick your poison; or do you have another idea?

mwc360 commented:
@keen85 thx! I'll review next Monday, I'm currently on PTO.

mwc360 commented:
@keen85 - connect with me on LinkedIn (www.linkedin.com/in/mileswcole) if you'd like and I'll mention you in the post announcing support for Sail, thx!
